lib/portlayer/store/store.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"sync"
+
+	"github.com/vmware/vic/pkg/kvstore"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/datastore"
+	"github.com/vmware/vic/pkg/vsphere/session"
+
+	"golang.org/x/net/context"
+)
+
+type StoreManager struct {
+	m            sync.RWMutex
+	dsStores     map[string]*kvstore.KeyValueStore
+	datastoreURL url.URL
+}
+
+var mgr *StoreManager
+
+const (
+	// default store folder name
+	dsFolder = "kvStores"
+	// available via portLayer API
+	APIKV = "apiKV"
+)
+
+var (
+	ErrDoesNotExit   = errors.New("requested store does not exist")
+	ErrDuplicateName = errors.New("duplicate store name")
+	ErrInvalidName   = errors.New("invalid store name, must be regexp(^[a-zA-Z0-9_]*$) compliant")
+)
+
+// Init will initialize the package vars and create the default portLayerKV store
+func Init(ctx context.Context, session *session.Session, imgStoreURL url.URL) error {
+	defer trace.End(trace.Begin(imgStoreURL.String()))
+
+	mgr = &StoreManager{
+		dsStores:     make(map[string]*kvstore.KeyValueStore),
+		datastoreURL: imgStoreURL,
+	}
+	//create or restore the api accessible datastore backed k/v store
+	return NewDatastoreKeyValue(ctx, session, APIKV)
+}
+
+// Store will return the requested store
+func Store(name string) (*kvstore.KeyValueStore, error) {
+	if _, exists := mgr.dsStores[name]; !exists {
+		return nil, ErrDoesNotExit
+	}
+	return mgr.dsStores[name], nil
+}
+
+// NewDatastoreKeyValue will validate the supplied name and create a datastore
+// backed key / value store
+//
+// The file will be located at the init StoresDatastoreURL  -- currently that's in the
+// appliance directory under the {dsFolder} folder (i.e. [datastore]vch-appliance/{dsFolder}/{name})
+func NewDatastoreKeyValue(ctx context.Context, session *session.Session, name string) error {
+	defer trace.End(trace.Begin(name))
+	// validate the name
+	err := validateStoreName(name)
+	if err != nil {
+		return err
+	}
+	// get a ds helper for this ds url
+	dsHelper, err := datastore.NewHelper(ctx, session, session.Datastore, fmt.Sprintf("%s/%s", mgr.datastoreURL.Path, dsFolder))
```

Comment:
question regarding need for future cleanup - if this is working now then it doesn't need changing immediately, but I don't think we formally ensure that session.Datastore is the datastore where the applianceVM lives.

This could be obtained by the following, although reading through it now I'm unsure what the state of the URL handling is. It's supposed to have some helpers to ensure that datastore and path pack into it and are extracted cleanly, but I cannot find them. @fdawg4l is this another tech debt area we need to address?:
```
ovm, _ := GetSelf() 
vm := vm.NewVirtualMachineFromVM(ctx, session, ovm)
ds := vm.DSPath.Host
```